sdks/java/core/src/main/java/org/apache/beam/sdk/io/TFRecordIO.java

```diff
@@ -0,0 +1,878 @@
+        checkNotNull(filepattern, "Filepattern cannot be empty.");
+        return new Bound(name, filepattern, validate, compressionType);
+      }
+
+      /**
+       * Returns a new transform for reading from TFRecord files that's like this one but
+       * that has GCS path validation on pipeline creation disabled.
+       *
+       * <p>This can be useful in the case where the GCS input does not
+       * exist at the pipeline creation time, but is expected to be
+       * available at execution time.
+       *
+       * <p>Does not modify this object.
+       */
+      public Bound withoutValidation() {
+        return new Bound(name, filepattern, false, compressionType);
+      }
+
+      /**
+       * Returns a new transform for reading from TFRecord files that's like this one but
+       * reads from input sources using the specified compression type.
+       *
+       * <p>If no compression type is specified, the default is
+       * {@link TFRecordIO.CompressionType#AUTO}.
+       * See {@link TFRecordIO.Read#withCompressionType} for more details.
+       *
+       * <p>Does not modify this object.
+       */
+      public Bound withCompressionType(TFRecordIO.CompressionType compressionType) {
+        return new Bound(name, filepattern, validate, compressionType);
+      }
+
+      @Override
+      public PCollection<byte[]> expand(PBegin input) {
+        if (filepattern == null) {
+          throw new IllegalStateException(
+              "need to set the filepattern of a TFRecordIO.Read transform");
+        }
+
+        if (validate) {
+          checkState(filepattern.isAccessible(), "Cannot validate with a RVP.");
+          try {
+            checkState(
+                !IOChannelUtils.getFactory(filepattern.get()).match(filepattern.get()).isEmpty(),
+                "Unable to find any files matching %s",
+                filepattern);
+          } catch (IOException e) {
+            throw new IllegalStateException(
+                String.format("Failed to validate %s", filepattern.get()), e);
+          }
+        }
+
+        final Bounded<byte[]> read = org.apache.beam.sdk.io.Read.from(getSource());
+        PCollection<byte[]> pcol = input.getPipeline().apply("Read", read);
+        // Honor the default output coder that would have been used by this PTransform.
+        pcol.setCoder(getDefaultOutputCoder());
+        return pcol;
+      }
+
+      // Helper to create a source specific to the requested compression type.
+      protected FileBasedSource<byte[]> getSource() {
+        switch (compressionType) {
+          case NONE:
+            return new TFRecordSource(filepattern);
+          case AUTO:
+            return CompressedSource.from(new TFRecordSource(filepattern));
+          case GZIP:
+            return
+                CompressedSource.from(new TFRecordSource(filepattern))
+                    .withDecompression(CompressedSource.CompressionMode.GZIP);
+          case DEFLATE:
```

Comment:
DELATE and ZLIB are essentially the same and the TensorFlow doc & API call it DEFLATE so it's more consistent with their choice.